# **Maestría en Inteligencia Artificial Aplicada**

## **Curso: Inteligencia Artificial y Aprendizaje Automático**

### Tecnológico de Monterrey

### Prof Luis Eduardo Falcón Morales

## **Actividad de las Semanas 5 y 6**
### **Problema de asignación de créditos: South German Dataset.**


## **Nombres y matrículas:**



*   ### Hiram García Austria: A00378771

*   ### Nombre y matrícula

*   ### Nombre y matrícula



# **Parte I: Partición, análisis y pre-procesamiento de los datos.**

## **Ejercicio 1**

In [63]:
# Aquí deberás incluir todas las librerías que requieras durante esta actividad:
import numpy as np
import pandas as pd

from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder, OrdinalEncoder
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from xgboost import XGBClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.svm import SVC
from sklearn.model_selection import RepeatedStratifiedKFold, train_test_split, cross_validate
from imblearn.metrics import geometric_mean_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from imblearn.over_sampling import SMOTE
from sklearn.metrics import make_scorer

# Bosque aleatorio
from sklearn.ensemble import RandomForestClassifier

In [64]:
# Si se desean comentar algunos de los Warnings.
import warnings
warnings.filterwarnings('ignore')

In [65]:
# Carga y renombra los nombres de las columnas del alemán al inglés y desplegamos
# de nuevo el DataFrame para ver el resultado obtenido:

# ************* Incluye aquí tu código:*****************************

df = pd.read_csv('./SouthGermanCredit.asc', sep=' ')
df.columns = [
    'status',
    'duration',
    'credit_history',
    'purpose',
    'amount',
    'savings',
    'employment_duration',
    'installment_rate',
    'personal_status_sex',
    'other_debtors',
    'present_residence',
    'property',
    'age',
    'other_installment_plans',
    'housing',
    'number_credits',
    'job',
    'people_liable',
    'telephone',
    'foreign_worker',
    'credit_risk'
]

# *********** Aquí termina la sección de agregar código *************

df.head().T

0     1    2     3     4
status                      1     1    2     1     1
duration                   18     9   12    12    12
credit_history              4     4    2     4     4
purpose                     2     0    9     0     0
amount                   1049  2799  841  2122  2171
savings                     1     1    2     1     1
employment_duration         2     3    4     3     3
installment_rate            4     2    2     3     4
personal_status_sex         2     3    2     3     3
other_debtors               1     1    1     1     1
present_residence           4     2    4     2     4
property                    2     1    1     1     2
age                        21    36   23    39    38
other_installment_plans     3     3    3     3     1
housing                     1     1    1     1     2
number_credits              1     2    1     2     2
job                         3     3    2     2     2
people_liable               2     1    2     1     2
telephone                   1     1    1     1     1
foreign_worker              2     2    2     1     1
credit_risk                 1     1    1     1     1

## **Ejercicio 2**

In [66]:
# Realiza a continuación una transformación para que la clase negativa (buen cliente)
# quede con el valor de 0 y la clase positiva (mal cliente) quede con el valor de 1.

# ************* Incluye aquí tu código:*****************************

df['credit_risk'] = df['credit_risk'].replace({'1': 0, '0': 1})

# *********** Aquí termina la sección de agregar código *************


print(df['credit_risk'].value_counts())

credit_risk
1    700
0    300
Name: count, dtype: int64


## **Ejercicio 3**

In [67]:
# Realiza una partición solicitada de entrenamiento y prueba.
# Los nombres de los conjuntos deberán ser como se indican en los print de abajo:

# ************* Incluye aquí tu código:*****************************

X = df.drop('credit_risk', axis=1)
y = df['credit_risk']

# Mismos porcentajes que se utilizaron en el artículo: 70% - 30%
Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)

# *********** Aquí termina la sección de agregar código *************


# Mostremos las dimensiones de la partición generada:
print("Dimensiones:")
print("Entrenamiento:", Xtrain.shape, ytrain.shape)
print("Prueba:", Xtest.shape, ytest.shape)

# Y el porcentaje de cada clase de la variable de salida:
tmp = ytrain.sum()/ytrain.shape[0]
print("\nPorcentaje clases Positiva:%.2f%%, y Negativa:%.2f%%" % (100*(1-tmp),tmp*100))


Dimensiones:
Entrenamiento: (700, 20) (700,)
Prueba: (300, 20) (300,)

Porcentaje clases Positiva:30.00%, y Negativa:70.00%


### **Con base al porcentaje de los niveles de la variable de salida ¿podemos decir que tenemos un problema de datos desbalanceado? ¿Por qué?**



### ++++++++ Inicia la sección de agregar texto: +++++++++++

Tenemos un problema de clases desbalanceadas ya que la clase positiva (mal cliente) es minoritaria (30%), mientras que la clase negativa (buen cliente) es mayoritaria (70%).

Para resolver este problema podemos realizar un balanceo de las clases en el conjunto de entrenamiento con técnicas de sobre-muestreo en la clase minoritaria, o bien técnicas de sub-muestreo en la clase mayoritaria.

### ++++++++ Termina la sección de agregar texto: +++++++++++



## **Ejercicio 4**

In [68]:
# De acuerdo a la información de la Tabla 3 del artículo de la IEEE
# define las variables correspondientes que se indican a continuación:

# ************* Incluye aquí tu código:*****************************

# Variables numéricas:
lista_paper_num = [
    'duration',
    'amount',
    'age',
    'people_liable'
]

# Variables ordinales:
lista_paper_ord = [
    'employment_duration',
    'installment_rate',
    'present_residence',
    'property',
    'number_credits',
    'job'
]

# Variables nominales:
lista_paper_cat = [
    'status',
    'credit_history',
    'purpose',
    'savings',
    'personal_status_sex',
    'other_debtors',
    'other_installment_plans',
    'housing',
    'telephone',
    'foreign_worker',
]


# *********** Aquí termina la sección de agregar código *************


## **Ejercicio 5**

In [69]:
# Transformaciones que se aplicarán a las variables numéricas usando la clase Pipeline de sklearn:

# ************* Incluye aquí tu código:*****************************

# Variables numéricas:
numericas_pipe = Pipeline([
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', MinMaxScaler())
])
numericas_pipe_nombres = lista_paper_num


# Variables categóricas-Nominales:
nominales_pipe = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore', drop='first'))
])
nominales_pipe_nombres = lista_paper_cat

# Variables categóricas-ordinales:
ordinales_pipe = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('ordinal', OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1))
])
ordinales_pipe_nombres = lista_paper_ord

# Conjuntas las transformaciones de todo tipo de variable y
# deja sin procesar aquellas que hayas decidido no transformar:

columnasTransformer = ColumnTransformer([
    ('numericas', numericas_pipe, numericas_pipe_nombres),
    ('nominales', nominales_pipe, nominales_pipe_nombres),
    ('ordinales', ordinales_pipe, ordinales_pipe_nombres)
], remainder='passthrough')

# *********** Aquí termina la sección de agregar código *************


## **Ejercicio 6**

In [70]:
# Como se va a utilizar Validación-Cruzada, concatena los conjuntos de entrenamiento y prueba
# en uno nuevo conjunto aumentado que llamaremos trainval para utilizar como entrenamiento:


# ************* Incluye aquí tu código:**************************

Xtraintest = pd.concat([Xtrain, Xtest], axis=0)
ytraintest = pd.concat([ytrain, ytest], axis=0)

# *********** Aquí termina la sección de agregar código *************


# Veamos cuántas variables nuevas se introducen con las transformaciones One-Hot-Encoding:
Xtmp = Xtraintest.copy()
tmp = columnasTransformer.fit_transform(Xtmp)
print("Dimensión de las variables de entrada ANTES de las transformaciones:", Xtmp.shape)
print("Dimensión de las variables de entrada DESPUÉS de las transformaciones:", tmp.shape)

Dimensión de las variables de entrada ANTES de las transformaciones: (1000, 20)
Dimensión de las variables de entrada DESPUÉS de las transformaciones: (1000, 41)


## **Ejercicio 7**

In [71]:
# Definimos a continuación la función que llamamos "mis_modelos" que incluye
# todos los modelos que deseamos comparar en el ejercicio.


def mis_modelos():
  modelos, nombres = list(), list()


  # ************* Incluye aquí tu código:**************************
  #
  # Deberás incluir en cada modelo los argumentos que consideres
  # adecuados para que cada uno converja y no esté sobre-entrenado
  # con respecto a la métrica de la exactitud (accuracy).


  # Regresión Logística - Logistic Regression-LR:
  # https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html

  # Este valor de "penalty" no se debe de cambiar, ya que define el modelo sin regularización.
  modelos.append(LogisticRegression(penalty=None, solver='lbfgs', max_iter=2000, random_state=1))
  nombres.append('LR')




  # k-Vecinos más Cercanos : k-Nearest-Neighbors-kNN:
  # https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html

  modelos.append(KNeighborsClassifier())
  nombres.append('kNN')




  # Árbol de decisiones-DecisionTree-DT:
  # https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html

  modelos.append(DecisionTreeClassifier(max_depth = None,       # None, 5, 3, 1
                                        min_samples_split=2,   # 2,3,5, 20
                                        #min_samples_leaf=1,  # trata al inicio de usar solo uno de estos, split o leaf, para su mejor comprensión.
                                        #random_state=7
                                        ))
  nombres.append('DTree')




  # Bosque Aleatorio-RandomForest-RF:
  # https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html

  modelos.append(RandomForestClassifier(n_estimators= 100,    # 100
                                        max_depth= None,      # None, 1, 2,3, 4, 5, 6 ... ¿Se esperaría la misma profunidad en un RF y en un DT?
                                        min_samples_split=2,    # 2, 5, 15
                                        #random_state=0
                                        ))
  nombres.append('RF')




  # XGBoosting:
  # https://xgboost.readthedocs.io/en/latest/python/python_api.html#xgboost.XGBClassifier
  # https://xgboost.readthedocs.io/en/stable/parameter.html

  modelos.append(XGBClassifier(booster= 'gbtree',
                               n_estimators=100,   # A medida que se aumente "n_estimators", se debe disminuir el "learning_rate", de manera general.
                               max_depth= 6,             # 1,3, 6,
                               learning_rate=0.3,   #  0.3, 0.000001, 0.01, 100,     # participación o peso de cada árbol desde el inicio.
                               subsample=1.0,        # 1.0,  0.9, 0.8, ... 0.5    # submuestreo con respecto a los renglones para evitar overfitting.
                               #random_state=5,
                               objective='binary:logistic',
                               n_jobs=-1))
  nombres.append('XGBoost')




  # Red neuronal de Perceptrón Multicapa-MLP:
  # https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html

  modelos.append(MLPClassifier(hidden_layer_sizes=(30,),     # 100,  6, 24, 30, (15,15), (50,50)
                               activation='logistic',
                               max_iter=1000,                  # 200,
                               alpha=0.0001,               # término de regularización L2.
                               #learning_rate='constant',       # tasa de aprendizaje o tamaño de paso del método Gradiente Descendente.
                               #learning_rate_init=0.001,
                               #random_state=1
                               ))
  nombres.append('MLP')




  # Máquina de Vectores de Soporte-SVM:
  # https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html

  modelos.append(SVC(kernel='rbf',
                     C= 1.0,      # 1.0, 0.00001, 1000           # inversamente proporcional a la constante de regularización L2.
                     gamma= 'scale',           # scale,  0.005
                     #class_weight='balanced',     # Siempre puedes hacer uso del balanceo en caso de que ayude.
                     #random_state=7
                     ))
  nombres.append('SVM')

  return modelos, nombres


# Técnica de sub muestreo (under sampling) y/o sobre muestreo (oversampling) utilizada:
mi_uoSampling = SMOTE(sampling_strategy=0.5, random_state=42) # Aplicando una relación de 50% de la clase minoritaria.

Xtv_uo, ytv_uo = mi_uoSampling.fit_resample(Xtraintest, ytraintest)



# *********** Aquí termina la sección de agregar código *******************





# Entrenamos cada uno de los modelos y desplegamos la métricas de Train y Val.

# NOTA: Observa que el método de Validación-Cruzada llama  a los resultados
#       de "validation" como "test":

modelos, nombres = mis_modelos()
resultados = list()

for i in range(len(modelos)):

  # Definimos nuestro pipeline con las transformaciones y los modelos:
  pipeline = Pipeline(steps=[('ct',columnasTransformer),('m',modelos[i])])

  # Aplicaremos validación-cruzada:
  micv = RepeatedStratifiedKFold(n_splits=5,
                                 n_repeats=3,
                                 random_state=5     # agreguemos una semilla para estabilizar resultados.
                                 )


  # Definimos las métricas que desamos recuperar:
  mismetricas = {'miaccuracy':'accuracy','miprecision':'precision','mirecall':'recall',
                 'mifi':'f1','miauc':'roc_auc','migmean':make_scorer(geometric_mean_score)}

  # Llevamos a cabo el entrenamiento:
  scores = cross_validate(pipeline,
                          Xtv_uo,
                          ytv_uo,
                          scoring=mismetricas,
                          cv=micv,
                          return_train_score=True,
                          )

  # Guardemos el resultado de cada modelo para análisis posteriores.
  resultados.append(scores)

  # Desplegamos los valores de las métricas para verificar si no hay
  # subentrenamiento o sobreentrenamiento:
  print('>> %s' % nombres[i])
  for j,k in enumerate(list(scores.keys())):
    if j>1:
      print('\t %s %.3f (%.3f)' % (k, np.mean(scores[k]),np.std(scores[k])))

>> LR
	 test_miaccuracy 0.752 (0.021)
	 train_miaccuracy 0.787 (0.007)
	 test_miprecision 0.797 (0.013)
	 train_miprecision 0.819 (0.007)
	 test_mirecall 0.842 (0.036)
	 train_mirecall 0.875 (0.005)
	 test_mifi 0.819 (0.018)
	 train_mifi 0.846 (0.005)
	 test_miauc 0.801 (0.023)
	 train_miauc 0.847 (0.007)
	 test_migmean 0.693 (0.022)
	 train_migmean 0.732 (0.012)
>> kNN
	 test_miaccuracy 0.706 (0.024)
	 train_miaccuracy 0.807 (0.007)
	 test_miprecision 0.757 (0.019)
	 train_miprecision 0.829 (0.008)
	 test_mirecall 0.826 (0.038)
	 train_mirecall 0.895 (0.008)
	 test_mifi 0.789 (0.019)
	 train_mifi 0.861 (0.005)
	 test_miauc 0.717 (0.030)
	 train_miauc 0.870 (0.008)
	 test_migmean 0.619 (0.041)
	 train_migmean 0.752 (0.012)
>> DTree
	 test_miaccuracy 0.687 (0.026)
	 train_miaccuracy 1.000 (0.000)
	 test_miprecision 0.768 (0.022)
	 train_miprecision 1.000 (0.000)
	 test_mirecall 0.761 (0.030)
	 train_mirecall 1.000 (0.000)
	 test_mifi 0.764 (0.020)
	 train_mifi 1.000 (0.000)
	 test_miauc

## **Ejercicio 8**

### **Escribe tus conclusiones finales de la actividad.**

### ++++++++ Inicia la sección de agregar texto: +++++++++++


## **Conclusiones sobre los resultados obtenidos**
* A pesar de que en la precisión, el mejor modelo es el Random Forest (RF) con 0.76, el modelo XGBoost tiene el segundo mejor desempeño con 0.756 (casi igual).
* En cuanto a la media geométrica, el modelo XGBoost tiene el mejor desempeño con 0.698.
* En cuanto al area bajo la curva (AUC), el modelo XGBoost tiene también el mejor desempeño con 0.809.
* En general XGBoost tiene el mejor desempeño en todas las métricas, por lo que consideramos que es el mejor modelo para este escenario.
* Como un pequeño detalle, respecto al tiempo de ejecución, el modelo XGBoost es el que más tiempo tarda en ejecutarse, pero consideramos que vale la pena por el desempeño que tiene.


### ++++++++ Termina la sección de agregar texto. +++++++++++


>> ### **Fin de la Actividad de las Semanas 5 y 6.**